File: cmd/ls.go
package cmd

import (
	"fmt"
	"sort"
	"strings"
	"time"

	"stash/store"

	"github.com/spf13/cobra"
)

func newLsCmd() *cobra.Command {
	var chars int
	var idMode string
	var metaFilters []string
	var dateMode string
	var sizeMode string
	var name bool
	var n int
	var preview bool
	var reverse bool
	var long bool

	cmd := &cobra.Command{
		Use:           "ls",
		Short:         "List stash entries",
		SilenceUsage:  true,
		SilenceErrors: true,
		RunE: func(c *cobra.Command, args []string) error {
			if dateMode != "" && dateMode != "absolute" && dateMode != "relative" && dateMode != "ls" {
				return fmt.Errorf("--date must be absolute, relative, or ls")
			}
			if sizeMode != "" && sizeMode != "human" && sizeMode != "bytes" {
				return fmt.Errorf("--size must be human or bytes")
			}
			if idMode != "short" && idMode != "full" && idMode != "pos" {
				return fmt.Errorf("--id must be short, full, or pos")
			}
			if len(args) > 0 {
				return fmt.Errorf("unexpected arguments: %s", strings.Join(args, " "))
			}
			effectiveDateMode := dateMode
			effectiveSizeMode := sizeMode
			effectiveName := name
			if long {
				if !c.Flags().Changed("date") {
					effectiveDateMode = "absolute"
				}
				if !c.Flags().Changed("size") {
					effectiveSizeMode = "human"
				}
				if !c.Flags().Changed("name") {
					effectiveName = true
				}
			}
			effectiveChars := chars

			metaSel, err := parseMetaSelection(metaFilters)
			if err != nil {
				return err
			}
			entries, err := collectEntries(metaSel, reverse, n)
			if err != nil {
				return err
			}

			now := time.Now()
			if preview && !c.Flags().Changed("chars") {
				effectiveChars = autoLSPreviewChars(entries, now, idMode, effectiveDateMode, effectiveSizeMode, effectiveName, metaSel)
			}
			return renderLS(entries, now, idMode, effectiveDateMode, effectiveSizeMode, effectiveName, preview, effectiveChars, metaSel)
		},
	}

	cmd.Flags().IntVar(&chars, "chars", 80, "Preview character limit")
	cmd.Flags().StringVar(&idMode, "id", "short", "ID display: short, full, or pos")
	cmd.Flags().StringArrayVarP(&metaFilters, "meta", "m", nil, "Show metadata tags with @, or filter by tag name (repeatable)")
	cmd.Flags().StringVar(&dateMode, "date", "", "Include date column: absolute, relative, or ls")
	cmd.Flags().Lookup("date").NoOptDefVal = "absolute"
	cmd.Flags().StringVar(&sizeMode, "size", "", "Include size column: human or bytes")
	cmd.Flags().Lookup("size").NoOptDefVal = "human"
	cmd.Flags().BoolVar(&name, "name", false, "Include filename or full ULID column")
	cmd.Flags().IntVarP(&n, "number", "n", 0, "Limit number of entries shown (0 = all)")
	cmd.Flags().BoolVarP(&preview, "preview", "p", false, "Append compact preview text")
	cmd.Flags().BoolVarP(&reverse, "reverse", "r", false, "Show oldest first")
	cmd.Flags().BoolVarP(&long, "long", "l", false, "Alias for --date --size --name")
	return cmd
}

func lsDate(t, now time.Time) string {
	if t.Year() == now.Year() {
		return t.Local().Format("Jan _2 15:04")
	}
	return t.Local().Format("Jan _2  2006")
}

func lsID(m store.Meta, idx int, idMode string) string {
	switch idMode {
	case "full":
		return m.DisplayID()
	case "pos":
		return fmt.Sprintf("%d", idx+1)
	default:
		return m.ShortID()
	}
}

func lsName(m store.Meta) string {
	if name := strings.TrimSpace(m.Attrs["filename"]); name != "" {
		return name
	}
	return m.DisplayID()
}

func lsAllAttrs(attrs map[string]string) string {
	if len(attrs) == 0 {
		return ""
	}
	keys := make([]string, 0, len(attrs))
	for k := range attrs {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	parts := make([]string, 0, len(keys))
	for _, k := range keys {
		parts = append(parts, attrs[k])
	}
	return strings.Join(parts, "  ")
}

type lsRow struct {
	id, name, metaInline, size, date, preview string
	metaVals                                  []string
}

func formatLSSize(size int64, mode string) string {
	if mode == "bytes" {
		return fmt.Sprintf("%d", size)
	}
	return store.HumanSize(size)
}

func lsMetaColumns(entries []store.Meta, sel metaSelection) []string {
	if len(sel.tags) > 0 {
		return append([]string(nil), sel.tags...)
	}
	return nil
}

func autoLSPreviewChars(entries []store.Meta, now time.Time, idMode, dateMode, sizeMode string, name bool, metaSel metaSelection) int {
	width, ok := terminalWidth()
	if !ok {
		return 80
	}
	metaCols := lsMetaColumns(entries, metaSel)
	metaWidths := make([]int, len(metaCols))
	maxID, maxSize, maxDate, maxName, maxInlineMeta := 0, 0, 0, 0, 0
	for i, m := range entries {
		if id := lsID(m, i, idMode); len(id) > maxID {
			maxID = len(id)
		}
		if sizeMode != "" {
			if size := formatLSSize(m.Size, sizeMode); len(size) > maxSize {
				maxSize = len(size)
			}
		}
		if dateMode != "" {
			if date := formatLSDate(parseTS(m.TS), now, dateMode); len(date) > maxDate {
				maxDate = len(date)
			}
		}
		if name {
			if n := lsName(m); len(n) > maxName {
				maxName = len(n)
			}
		}
		if len(metaCols) > 0 {
			for idx, col := range metaCols {
				value := " "
				if v, ok := m.Attrs[col]; ok {
					value = v
				}
				if len(value) > metaWidths[idx] {
					metaWidths[idx] = len(value)
				}
			}
		} else if metaSel.showAll {
			if inline := lsAllAttrs(m.Attrs); len(inline) > maxInlineMeta {
				maxInlineMeta = len(inline)
			}
		}
	}
	fixed := maxID
	if sizeMode != "" {
		fixed += 2 + maxSize
	}
	if dateMode != "" {
		fixed += 2 + maxDate
	}
	if name {
		fixed += 2 + maxName
	}
	for _, w := range metaWidths {
		fixed += 2 + w
	}
	if maxInlineMeta > 0 {
		fixed += 2 + maxInlineMeta
	}
	chars := width - fixed - 2
	if chars < 20 {
		return 20
	}
	return chars
}

func buildLSRows(entries []store.Meta, now time.Time, dateMode, sizeMode, idMode string, name, preview bool, chars int, metaSel metaSelection) []lsRow {
	metaCols := lsMetaColumns(entries, metaSel)
	rows := make([]lsRow, len(entries))
	for i, m := range entries {
		r := lsRow{
			id: lsID(m, i, idMode),
		}
		if len(metaCols) > 0 {
			r.metaVals = make([]string, len(metaCols))
			for idx, col := range metaCols {
				r.metaVals[idx] = " "
				if v, ok := m.Attrs[col]; ok {
					r.metaVals[idx] = v
				}
			}
		} else if metaSel.showAll {
			r.metaInline = lsAllAttrs(m.Attrs)
		}
		if name {
			r.name = lsName(m)
		}
		if sizeMode != "" {
			r.size = formatLSSize(m.Size, sizeMode)
		}
		if dateMode != "" {
			r.date = formatLSDate(parseTS(m.TS), now, dateMode)
		}
		if preview && m.Preview != "" {
			r.preview = previewSnippet(m.Preview, chars)
		}
		rows[i] = r
	}
	return rows
}

func renderLS(entries []store.Meta, now time.Time, idMode, dateMode, sizeMode string, name, preview bool, chars int, metaSel metaSelection) error {
	showMeta := metaSel.showAll || len(metaSel.tags) > 0
	if dateMode == "" && sizeMode == "" && !name && !preview && !showMeta {
		for i, m := range entries {
			fmt.Println(lsID(m, i, idMode))
		}
		return nil
	}

	rows := buildLSRows(entries, now, dateMode, sizeMode, idMode, name, preview, chars, metaSel)
	metaCols := lsMetaColumns(entries, metaSel)
	if dateMode == "ls" {
		for i, m := range entries {
			rows[i].date = lsDate(parseTS(m.TS), now)
		}
	}

	maxID, maxName, maxSize, maxDate, maxInlineMeta := 0, 0, 0, 0, 0
	metaWidths := make([]int, len(metaCols))
	for _, r := range rows {
		if len(r.id) > maxID {
			maxID = len(r.id)
		}
		if len(r.name) > maxName {
			maxName = len(r.name)
		}
		if len(r.size) > maxSize {
			maxSize = len(r.size)
		}
		if len(r.date) > maxDate {
			maxDate = len(r.date)
		}
		if len(r.metaInline) > maxInlineMeta {
			maxInlineMeta = len(r.metaInline)
		}
		for idx, v := range r.metaVals {
			if len(v) > metaWidths[idx] {
				metaWidths[idx] = len(v)
			}
		}
	}

	for _, r := range rows {
		idCol := clrID(fmt.Sprintf("%-*s", maxID, r.id))
		parts := []string{idCol}
		if r.size != "" {
			parts = append(parts, fmt.Sprintf("%*s", maxSize, r.size))
		}
		if r.date != "" {
			parts = append(parts, fmt.Sprintf("%*s", maxDate, r.date))
		}
		if r.name != "" {
			nameCol := r.name
			if nameCol != r.id {
				nameCol = clrFile(fmt.Sprintf("%-*s", maxName, r.name))
			} else {
				nameCol = fmt.Sprintf("%-*s", maxName, r.name)
			}
			parts = append(parts, nameCol)
		}
		for idx, v := range r.metaVals {
			parts = append(parts, clrAttrs(fmt.Sprintf("%-*s", metaWidths[idx], v)))
		}
		if r.metaInline != "" {
			parts = append(parts, clrAttrs(fmt.Sprintf("%-*s", maxInlineMeta, r.metaInline)))
		}
		if r.preview != "" {
			parts = append(parts, r.preview)
		}
		line := strings.Join(parts, "  ")
		if width, ok := terminalWidth(); ok {
			line = trimANSIToWidth(line, width)
		}
		fmt.Println(line)
	}
	return nil
}

File: src/cmd/ls.rs
use clap::{ArgAction, Args};
use std::io::{self, Write};

use crate::display::*;
use crate::store::parse_meta_selection;

#[derive(Args, Debug, Clone)]
pub(crate) struct LsArgs {
    #[arg(
        long,
        default_value = "short",
        help = "ID display: short, full, or pos"
    )]
    id: String,

    #[arg(short = 'a', long = "attr", value_name = "name|+name", action = ArgAction::Append, help = "Show an attribute column, or filter with +name (repeatable)")]
    attr: Vec<String>,

    #[arg(
        short = 'A',
        long = "attrs",
        num_args = 0..=1,
        default_missing_value = "list",
        value_name = "list|count|flag",
        help = "Attribute display: list, count, or flag"
    )]
    attrs: Option<String>,

    #[arg(
        short = 'n',
        long = "number",
        default_value_t = 0,
        help = "Limit number of entries shown (0 = all)"
    )]
    number: usize,

    #[arg(short = 'r', long = "reverse", help = "Show oldest first")]
    reverse: bool,

    #[arg(long, help = "Output listing as rich JSON")]
    json: bool,

    #[arg(long, default_missing_value = "ls", num_args = 0..=1, help = "Include date column: iso, ago, or ls")]
    date: Option<String>,

    #[arg(long, default_missing_value = "human", num_args = 0..=1, help = "Include size column: human or bytes")]
    size: Option<String>,

    #[arg(
        long,
        help = "Include filename (attribute) if available, or else full ULID column"
    )]
    name: bool,

    #[arg(short = 'p', long = "preview", help = "Append compact preview text")]
    preview: bool,

    #[arg(short = 'l', long = "long", help = "Alias for --date --size --name")]
    long: bool,

    #[arg(long, default_value_t = 80, help = "Preview character limit")]
    chars: usize,

    #[arg(long, default_value = "true", help = "Color output: true or false")]
    color: String,
}

#[derive(Copy, Clone, Debug, Eq, PartialEq)]
enum AttrsMode {
    List,
    Count,
    Flag,
}

fn parse_attrs_mode(value: Option<&str>) -> io::Result<Option<AttrsMode>> {
    match value {
        None => Ok(None),
        Some("list") => Ok(Some(AttrsMode::List)),
        Some("count") => Ok(Some(AttrsMode::Count)),
        Some("flag") => Ok(Some(AttrsMode::Flag)),
        Some(_) => Err(io::Error::new(
            io::ErrorKind::InvalidInput,
            "--attrs must be list, count, or flag",
        )),
    }
}

pub(super) fn ls_command(mut args: LsArgs) -> io::Result<()> {
    if args.long {
        args.date.get_or_insert("ls".into());
        args.size.get_or_insert("human".into());
        if args.attrs.is_none() {
            args.attrs = Some("flag".into());
        }
        if !args.attr.iter().any(|value| value == "filename") {
            args.attr.push("filename".into());
        }
    }
    let color = color_enabled(&args.color)?;
    let attrs_mode = parse_attrs_mode(args.attrs.as_deref())?;
    if let Some(mode) = args.date.as_deref() {
        args.date = Some(normalize_date_mode(mode)?.to_string());
    }
    let meta_sel = parse_meta_selection(&args.attr, attrs_mode == Some(AttrsMode::List))?;
    let items = super::collect_entries(&meta_sel, args.reverse, args.number)?;
    let ls_date_mode = args.date.as_deref().unwrap_or("ls");
    if args.json {
        print_entries_json(&items, ls_date_mode, args.chars);
        return Ok(());
    }
    if args.date.is_none()
        && args.size.is_none()
        && !args.name
        && !args.preview
        && attrs_mode != Some(AttrsMode::Count)
        && attrs_mode != Some(AttrsMode::Flag)
        && !meta_sel.show_all
        && meta_sel.display_tags.is_empty()
    {
        let stdout = io::stdout();
        let mut out = stdout.lock();
        let rows = decorate_entries(&items, &args.id, ls_date_mode, args.chars, &meta_sel);
        for row in rows {
            write_colored(&mut out, &row.id, "1;33", color)?;
            writeln!(out)?;
        }
        return Ok(());
    }

    struct LsRow {
        id: String,
        size: String,
        date: String,
        name: String,
        attr_count: String,
        attr_flag: String,
        meta_vals: Vec<String>,
        meta_inline: String,
        preview: String,
    }

    // When auto-computing preview width, build entries with empty previews first;
    // effective_chars will be derived from the column measurement loop below.
    let initial_chars = if args.preview && args.chars == 80 { 0 } else { args.chars };
    let decorated = decorate_entries(&items, &args.id, ls_date_mode, initial_chars, &meta_sel);
    let mut rows = Vec::with_capacity(decorated.len());
    for (row, item) in decorated.into_iter().zip(items.iter()) {
        let DecoratedEntry {
            id,
            size_bytes,
            size_human,
            date,
            preview,
            filename,
            meta_vals,
            meta_inline,
            log_attr_lines: _,
        } = row;
        rows.push(LsRow {
            id: id.clone(),
            size: args
                .size
                .as_deref()
                .map(|mode| {
                    if mode == "bytes" {
                        size_bytes.clone()
                    } else {
                        size_human.clone()
                    }
                })
                .unwrap_or_default(),
            date: if args.date.is_some() {
                date
            } else {
                String::new()
            },
            name: if args.name {
                filename.unwrap_or_else(|| id.clone())
            } else {
                String::new()
            },
            attr_count: if attrs_mode == Some(AttrsMode::Count) {
                item.attrs.len().to_string()
            } else {
                String::new()
            },
            attr_flag: if attrs_mode == Some(AttrsMode::Flag) && !item.attrs.is_empty() {
                "*".to_string()
            } else {
                String::new()
            },
            meta_vals,
            meta_inline,
            preview: if args.preview { preview } else { String::new() },
        });
    }

    let mut max_id = 0usize;
    let mut max_size = 0usize;
    let mut max_date = 0usize;
    let mut max_name = 0usize;
    let mut max_attr_count = 0usize;
    let mut max_attr_flag = 0usize;
    let mut max_inline_meta = 0usize;
    let mut meta_widths = vec![0usize; meta_sel.display_tags.len()];
    for row in &rows {
        max_id = max_id.max(row.id.chars().count());
        max_size = max_size.max(row.size.chars().count());
        max_date = max_date.max(row.date.chars().count());
        max_name = max_name.max(row.name.chars().count());
        max_attr_count = max_attr_count.max(row.attr_count.chars().count());
        max_attr_flag = max_attr_flag.max(row.attr_flag.chars().count());
        max_inline_meta = max_inline_meta.max(row.meta_inline.chars().count());
        for (idx, value) in row.meta_vals.iter().enumerate() {
            meta_widths[idx] = meta_widths[idx].max(value.chars().count());
        }
    }

    let width = terminal_width();
    if args.preview && args.chars == 80 {
        let term_width = width.unwrap_or(80);
        let mut fixed = max_id;
        if max_size > 0 { fixed += 2 + max_size; }
        if max_date > 0 { fixed += 2 + max_date; }
        if max_name > 0 { fixed += 2 + max_name; }
        if max_attr_count > 0 { fixed += 2 + max_attr_count; }
        if max_attr_flag > 0 { fixed += 2 + max_attr_flag; }
        for &mw in &meta_widths { fixed += 2 + mw; }
        if max_inline_meta > 0 { fixed += 2 + max_inline_meta; }
        let effective_chars = term_width.saturating_sub(fixed + 2).max(20);
        for (row, item) in rows.iter_mut().zip(items.iter()) {
            if !item.preview.is_empty() {
                row.preview = preview_snippet(&item.preview, effective_chars);
            }
        }
    }

    let stdout = io::stdout();
    let mut out = stdout.lock();
    for row in rows {
        let mut line = String::new();
        push_colorized(&mut line, &pad_right(&row.id, max_id), "1;33", color);
        if !row.size.is_empty() {
            line.push_str("  ");
            line.push_str(&pad_left(&row.size, max_size));
        }
        if !row.date.is_empty() {
            line.push_str("  ");
            line.push_str(&pad_left(&row.date, max_date));
        }
        if !row.name.is_empty() {
            line.push_str("  ");
            let padded = pad_right(&row.name, max_name);
            if row.name == row.id {
                line.push_str(&padded);
            } else {
                push_colorized(&mut line, &padded, "1;36", color);
            }
        }
        if !row.attr_count.is_empty() {
            line.push_str("  ");
            push_colorized(&mut line, &pad_left(&row.attr_count, max_attr_count), "35", color);
        }
        if max_attr_flag > 0 {
            line.push_str("  ");
            push_colorized(&mut line, &pad_left(&row.attr_flag, max_attr_flag), "35", color);
        }
        for (idx, value) in row.meta_vals.iter().enumerate() {
            line.push_str("  ");
            push_colorized(&mut line, &pad_right(value, meta_widths[idx]), "35", color);
        }
        if !row.meta_inline.is_empty() {
            line.push_str("  ");
            push_colorized(
                &mut line,
                &pad_right(&row.meta_inline, max_inline_meta),
                "35",
                color,
            );
        }
        if !row.preview.is_empty() {
            line.push_str("  ");
            line.push_str(&row.preview);
        }
        let rendered = if let Some(width) = width {
            trim_ansi_to_width(&line, width)
        } else {
            line
        };
        writeln!(out, "{rendered}")?;
    }
    Ok(())
}
